pkg/controller/replicaset/replica_set.go

```diff
@@ -794,14 +794,19 @@ func (rsc *ReplicaSetController) syncReplicaSet(ctx context.Context, key string)
 	if manageReplicasErr != nil {
 		return manageReplicasErr
 	}
-	// Resync the ReplicaSet after MinReadySeconds as a last line of defense to guard against clock-skew.
+	// Plan the next availability check as a last line of defense against queue preemption and guard against clock-skew.
 	if updatedRS.Spec.MinReadySeconds > 0 &&
 		updatedRS.Status.ReadyReplicas == *(updatedRS.Spec.Replicas) &&
 		updatedRS.Status.AvailableReplicas != *(updatedRS.Spec.Replicas) {
-		nextSyncInSeconds = ptr.To(int(updatedRS.Spec.MinReadySeconds))
+		if nextCheck := controller.FindMinNextPodAvailabilityCheck(activePods, updatedRS.Spec.MinReadySeconds, rsc.clock.Now()); nextCheck != nil {
+			nextSyncDuration = ptr.To(*nextCheck + time.Second) // Add a second to avoid a clock-skew.
```

Comment:
Adding a second does not feel like a reliable strategy, and may cause unnecessary delays, or subtle bugs if the code is not ready for skew of >1s. 

Is it a common practice in other controllers? For example, we don't use this tactic in Job controller when handling activeDeadlineSeconds, see [here](https://github.com/kubernetes/kubernetes/blob/cc1fc3eea0e3e2b2674b6dd1cd6db01863088eee/pkg/controller/job/job_controller.go#L985-L987).

IIUC, even if the clock is skewed, we should just schedule another sync. If this is the case, I would prefer to avoid the extra 1s added.

Reply:
After some testing, I think we can remove the 1s, and try to resync instead if the availability is not reached.

Btw, there are actually two skews:

- https://github.com/kubernetes/kubernetes/issues/39785#issuecomment-279959133 which I think we will fix with the repeating resyncs
- https://github.com/kubernetes/kubernetes/issues/29229 kubelet/master skew which I am not sure we can easily fix as far as I can see